--- mesonbuild/modules/gnome.py.glib	2017-03-31 16:49:23.610293792 +0100
+++ mesonbuild/modules/gnome.py	2017-03-31 16:49:35.033272261 +0100
@@ -37,7 +37,7 @@ from . import ExtensionModule
 #
 # https://github.com/ninja-build/ninja/issues/1184
 # https://bugzilla.gnome.org/show_bug.cgi?id=774368
-gresource_dep_needed_version = '>= 2.51.1'
+gresource_dep_needed_version = '>= 2.50.0'
 
 native_glib_version = None
 girwarning_printed = False
